components/locid/src/subtags/variant.rs

```diff
@@ -2,209 +2,32 @@
-    /// representing a [`Variant`] subtag in canonical syntax.
-    pub const unsafe fn from_raw_unchecked(v: [u8; 8]) -> Self {
-        Self(TinyAsciiStr::from_bytes_unchecked(v))
-    }
-
-    /// A helper function for displaying
-    /// a [`Variant`] subtag as a `&`[`str`].
+    /// [`Variant`] represents a Unicode base language code conformat to the
+    /// [`unicode_variant_id`] field of the Language and Locale Identifier.
     ///
     /// # Examples
     ///
     /// ```
     /// use icu::locid::subtags::Variant;
     ///
-    /// let variant = Variant::from_bytes(b"macos").expect("Parsing failed.");
-    ///
-    /// assert_eq!(variant.as_str(), "macos");
-    /// ```
-    ///
-    /// `Notice`: For many use cases, such as comparison,
-    /// [`Variant`] implements [`PartialEq`]`<&`[`str`]`>` which allows for direct comparisons.
-    pub fn as_str(&self) -> &str {
-        self.0.as_str()
-    }
-
-    /// Compare this `Variant` with BCP-47 bytes.
-    ///
-    /// The return value is equivalent to what would happen if you first converted this
-    /// `Variant` to a BCP-47 string and then performed a byte comparison.
-    ///
-    /// This function is case-sensitive and results in a *total order*, so it is appropriate for
-    /// binary search. The only argument producing [`Ordering::Equal`](core::cmp::Ordering::Equal)
-    /// is `self.to_string()`.
-    #[inline]
-    pub fn strict_cmp(&self, other: &[u8]) -> core::cmp::Ordering {
-        self.as_str().as_bytes().cmp(other)
-    }
-
-    /// Compare this `Variant` with a potentially unnormalized BCP-47 string.
-    ///
-    /// The return value is equivalent to what would happen if you first parsed the
-    /// BCP-47 string to a `Variant` and then performed a structucal comparison.
-    ///
-    #[inline]
-    pub fn normalizing_eq(&self, other: &str) -> bool {
-        self.as_str().eq_ignore_ascii_case(other)
-    }
-}
-
-impl FromStr for Variant {
-    type Err = ParserError;
-
-    fn from_str(source: &str) -> Result<Self, Self::Err> {
-        Self::from_bytes(source.as_bytes())
-    }
-}
-
-impl_writeable_for_single_subtag!(Variant, "posix");
-
-impl<'l> From<&'l Variant> for &'l str {
-    fn from(input: &'l Variant) -> Self {
-        input.as_str()
-    }
-}
-
-impl From<Variant> for TinyAsciiStr<8> {
-    fn from(input: Variant) -> Self {
-        input.0
-    }
-}
+    /// let variant: Variant = "macos".parse().expect("Failed to parse a variant subtag.");
+    /// ```
+    ///
+    /// [`unicode_variant_id`]: https://unicode.org/reports/tr35/#unicode_variant_id
+    Variant,
+    subtags::Variant,
+    subtags_variant,
+    4..=8,
+    s,
+    s.is_ascii_alphanumeric() && (s.len() != 4 || s.all_bytes()[0].is_ascii_digit()),
+    s.to_ascii_lowercase(),
+    s.is_ascii_lowercase()
+        && s.is_ascii_alphanumeric()
+        && (s.len() != 4 || s.all_bytes()[0].is_ascii_digit()),
```

Comment:
Praise: This is a clearer way to express the invariant than we had before

```
 variant       = 5*8alphanum         ; registered variants
               / (DIGIT 3alphanum)
```

https://www.rfc-editor.org/rfc/bcp/bcp47.txt